### ISSUE TITLE
How to explain the the results of evalPerPatch and evalPerImage?

### ISSUE THREAD

**Issue by dearhoper** [facebookresearch/deepmask#75]
Who can tell me the detailed information about the evaluation results? (under the bold font)
Is 41 the average speed of samples analysis? How many samples? One batch? 32?
And what about the mean and median results? ARs?

`$ sudo th evalPerPatch.lua exps/deepmask/exp/ -save`
`| loading model file...exps/deepmask/exp/	`
`| number of examples seen until now: 9344000 (73 epochs)	`
`| start per batch evaluation	`
**`[=================== 200/200 ====================>]  Tot: 2m34s | Step: 726ms`**
`| finish	`
`----------------------------------------------	`
`| model: exps/deepmask/exp/`
`| # epochs: 73`
`| # samples: 6400`
**`| samples/s      41 | mean 065.47 median 073.20 iou@.5 077.06  iou@.7 055.15`** 

`$ sudo th evalPerImage.lua exps/deepmask/exp/ -save`
`| loading model file... exps/deepmask/exp/	`
`| number of examples seen until now: 9344000 (73 epochs)	`
`| saving results results in exps/deepmask/exp//epoch=73/	`
`| start eval	`
**`[================= 5000/5000 ================>]  Tot: 2h56m | Step: 2s86ms`**   
`| finish	`

**Comment by PavlosMelissinos** [comment]
1. [This is the line you're interested in](https://github.com/facebookresearch/deepmask/blob/master/evalPerPatch.lua#L155)

> config.batch*config.testmaxload/sys.toc()

As you can see, it's the total number of images processed over the time spent to do so.

2. [maskMeter is a IoU metric](https://github.com/facebookresearch/deepmask/blob/master/evalPerPatch.lua#L77), therefore mean is mean IoU and median is median IoU.

[See this part of the code for more information](https://github.com/facebookresearch/deepmask/blob/master/trainMeters.lua#L90-L112).

**Comment by dearhoper** [comment]
Yes, it's right. 32*200/154s = 41
But, calculated in another way: 154s/6400=0.024s per image ?
It's not consistent with the description in the paper:

> Speed: Inference takes an average of 1.6s per image in the COCO dataset (1.2s on the smaller PASCAL images).

**Comment by PavlosMelissinos** [comment]
You're right, nice observation!

I should have read the code more carefully because samples (in evalPerPatch) refer to patches, not images (each image is loaded once but there are multiple patches in it), like I said. That means each sample is smaller (in number of pixels) than a full image, therefore processing it takes less time.

Moreover, [saveRes](https://github.com/facebookresearch/deepmask/blob/master/evalPerPatch.lua#L81) is called once per batch and outputs a single image.

Therefore, if I'm reading this correctly, every sample in a batch corresponds to a single image, so in your case it's actually more like config.testmaxload (200) images and config.batch (32) batches/patches/proposals per image.

That means time per image is about 770ms. This is still less than the reported one but as you can see on the [homepage of the project](https://github.com/facebookresearch/deepmask#evaluation):

> There are two ways to evaluate a model on the COCO dataset.
> 
> 1. evalPerPatch.lua **evaluates only the mask generation step**. The per-patch evaluation only uses image patches that contain roughly centered objects. Its usage is as follows:
> 
>    `th evalPerPatch.lua /path/to/trained/deepmask-or-sharpmask/`
> 2. evalPerImage.lua **evaluates the full model on COCO images, as reported in the papers**. By default, it evaluates performance on the first 5K COCO validation images (run th evalPerImage.lua --help to see the options):
> 
>     `th evalPerImage.lua /path/to/trained/deepmask-or-sharpmask/`

which I believe should answer your question.

**Comment by dearhoper** [comment]
Your're right!  0.77s per image is just the mask generation speed, not all of the time spent.
According to the content you quoted, I should calculate the true speed through the evalPerImage commad.

`[================= 5000/5000 ================>] Tot: 2h56m | Step: 2s86ms`

(2*60+56)*60/5000 = 2.112s per image, is it right?

 Thank you for the reply!

**Comment by PavlosMelissinos** [comment]
Yes, that looks alright.

Btw, if your problem has been solved please close the issue. :)

**Comment by chenyanlu** [comment]
hello,i have a problem when i run "th evalPerPatch.lua ./pretrained/sharpmask/"
i need help,thank you.
| loading model file..../pretrained/sharpmask/
/home/vr/torch/install/bin/luajit: /home/vr/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] /home/vr/torch/install/share/lua/5.1/coco/CocoApi.lua:126: assertion failed!
stack traceback:
[C]: in function 'assert'
/home/vr/torch/install/share/lua/5.1/coco/CocoApi.lua:126: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'CocoApi'
/home/vr/deepmask/DataSampler.lua:24: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'DataSampler'
/home/vr/deepmask/DataLoader.lua:37: in function </home/vr/deepmask/DataLoader.lua:32>
[C]: in function 'xpcall'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
/home/vr/torch/install/share/lua/5.1/threads/queue.lua:65: in function </home/vr/torch/install/share/lua/5.1/threads/queue.lua:41>
[C]: in function 'pcall'
/home/vr/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
[string " local Queue = require 'threads.queue'..."]:13: in main chunk
stack traceback:
[C]: in function 'error'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:264: in function 'synchronize'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:142: in function 'specific'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:125: in function 'Threads'
/home/vr/deepmask/DataLoader.lua:41: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'DataLoader'
/home/vr/deepmask/DataLoader.lua:21: in function 'create'
evalPerPatch.lua:77: in main chunk
[C]: in function 'dofile'
...e/vr/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
[C]: at 0x00405d50

**Comment by chenyanlu** [comment]
@dearhoper 
hello,i have a problem when i run "th evalPerPatch.lua ./pretrained/sharpmask/"
i need help,thank you.

| loading model file..../pretrained/sharpmask/
/home/vr/torch/install/bin/luajit: /home/vr/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] /home/vr/torch/install/share/lua/5.1/coco/CocoApi.lua:126: assertion failed!
stack traceback:
[C]: in function 'assert'
/home/vr/torch/install/share/lua/5.1/coco/CocoApi.lua:126: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'CocoApi'
/home/vr/deepmask/DataSampler.lua:24: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'DataSampler'
/home/vr/deepmask/DataLoader.lua:37: in function </home/vr/deepmask/DataLoader.lua:32>
[C]: in function 'xpcall'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
/home/vr/torch/install/share/lua/5.1/threads/queue.lua:65: in function </home/vr/torch/install/share/lua/5.1/threads/queue.lua:41>
[C]: in function 'pcall'
/home/vr/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
[string " local Queue = require 'threads.queue'..."]:13: in main chunk
stack traceback:
[C]: in function 'error'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:264: in function 'synchronize'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:142: in function 'specific'
/home/vr/torch/install/share/lua/5.1/threads/threads.lua:125: in function 'Threads'
/home/vr/deepmask/DataLoader.lua:41: in function '__init'
/home/vr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/vr/torch/install/share/lua/5.1/torch/init.lua:87>
[C]: in function 'DataLoader'
/home/vr/deepmask/DataLoader.lua:21: in function 'create'
evalPerPatch.lua:77: in main chunk
[C]: in function 'dofile'
...e/vr/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
[C]: at 0x00405d50

**Comment by dearhoper** [comment]
@chenyanlu 
The 126th line of CocoApi.lua is:
`assert( string.sub(annFile,-4,-1)=='json' and paths.filep(annFile) )`
Maybe you should make sure that the instances_val2014.json file exists, under the data/ folder.

**Comment by dearhoper** [comment]
Aha, you're stepping into the same road, repeating the same question I met 2 months ago.
The reason is : `luajit has a heap limit of 2GB in lua-land`
Please refer: [Failed to run "th train.lua" ... #50](https://github.com/facebookresearch/deepmask/issues/50)

**Comment by chenyanlu** [comment]
Oh, Thank you, @dearhoper 
I  rebuild torch using lua 5.2 instead of LuaJIT, then, getting an error .
At first , I think packages are changed. 
So, I update some packages.(argchech   coco   cudnn  cutorch  cunn  cwrap   dok   env   gnuplot   graph   image   inn   json  lua-cjson  luaffi    luafilesystem    nn  nngraph   nnx   optim   paths  penlight   qtlua   qttorch   sundown   sys   tds   threads   torch   trepl   xlua)
However,stil that eror.
What should I do?


th> require 'cunn'
not found: THNN_CudaHalfELU_updateOutput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type	
not found: THNN_CudaHalfELU_updateGradInput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type	
not found: THNN_CudaHalfHardTanh_updateOutput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type	
not found: THNN_CudaHalfHardTanh_updateGradInput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type	
not found: THNN_CudaHalfLeakyReLU_updateOutput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type

**Comment by dearhoper** [comment]
@chenyanlu 
Just warnings. Don't worry.
And someone is attempting to fix it. See [THNN_CudaHalf not found #380](https://github.com/torch/cunn/issues/380).

**Comment by chenyanlu** [comment]
Hum, thanks  !  @dearhoper 
When this is LuaJIT , I can run successfully.
I just run './clean.sh' and 'TORCH_LUA_VERSION=LUA52 ./install.sh' to instead of  LuaJIT .
I run ' th computeProposals.lua pretrained/deepmask  ', it have this error.
I did right?

**Comment by dearhoper** [comment]
@chenyanlu 
Yes, it's right.

**Comment by chenyanlu** [comment]
Hum, @dearhoper .
I rebuild torch using lua 5.2 instead of LuaJIT.
Then,I run ' th computeProposals.lua pretrained/deepmask ',it have this error.
how to solve this error?

not found: THNN_CudaHalfELU_updateOutput/home/vr/torch/install/share/lua/5.2/nn/THNN.lua:108: NYI: call arg type

**Comment by dearhoper** [comment]
@chenyanlu 
It's just a warning, not an error. Ignor it.
